rustls/src/suites.rs

```diff
@@ -181,6 +195,48 @@ pub(crate) fn reduce_given_version_and_protocol(
         .collect()
 }
 
+/// Return a list of the ciphersuites in `all` with the suites
+/// incompatible with the Groups extension removed.
+pub(crate) fn reduce_given_kx_groups(
+    all: &[SupportedCipherSuite],
+    groups_ext: Option<&[NamedGroup]>,
+    supported_groups: &[NamedGroup],
+) -> Vec<SupportedCipherSuite> {
+    fn is_ffdhe(g: &NamedGroup) -> bool {
+        match g {
+            // Codepoints between 0x100 and 0x200 are reserved for ffdhe groups.
+            // https://datatracker.ietf.org/doc/html/rfc7919#section-2
+            NamedGroup::Unknown(x) => (0x100..0x200).contains(x),
+            _ => g.key_exchange_algorithm() == Some(KeyExchangeAlgorithm::DHE),
+        }
+    }
+    // https://datatracker.ietf.org/doc/html/rfc7919#section-4 (paragraph 1)
+    let no_known_ffdhe_groups = if let Some(groups_ext) = groups_ext {
+        let mut ext_has_ffdhe_groups = false;
+        let mut ext_has_known_ffdhe_groups = false;
+        for g in groups_ext
+            .iter()
+            .filter(|g| is_ffdhe(g))
+        {
```

Comment:
Would prefer to avoid the `filter()` combinator in the loop header in favor of inlining the `match` from `is_ffdhe()` in the loop body with an empty return.